toc.md

```diff
@@ -428,77 +429,232 @@ An additional backlog of 936 hours of image processing jobs everyday, along with
+We can see that having more processes translates into increased performance. In each of these cases, time is divided by two. This is not a 100% guarantee, some factors need to be taken into account, such as:
+- other processes running and requiring some computational resources, effectively reducing the CPU time scheduled for our processes
+- relying on the OS scheduler for optimizing the sharing of computational resources, effectively introducing randomness from our point of view
+
+In the next section we will see how we introduced parallelism specifically for our BJP Workerholic.
 
 ##### Parallelism in Workerholic
+
+The diagram below shows how Workerholic uses multiple processes if we had the CPU cores available to do so. The OS scheduler pushes / evicts processes from the CPU and each process has its own worker threads that can poll jobs from Redis and work on them. So if we have four CPU cores we can have computational resources allocated to four different processes potentially at the same time, effectively allowing them to run in parallel.
+
 ![parallelism_workeholic_diagram](/images/parallelism_workeholic_diagram.png)
 
-The diagram above shows how Workerholic uses multiple processes if we had the CPU cores available to do so. The OS scheduler schedules the cores to each process, each process has its own worker threads which can poll jobs from Redis and work on them. So if we have four CPU cores we can have computational resources allocated to four different processes potentially at the same time, effectively allowing them to run in parallel, reducing the enqueuing:processing ratio from 1:40 to 1:10.
+Here is the code we wrote to make the above model possible:
+
+```ruby
+module Workerholic
+  class Starter
+    # ...
+
+    def self.launch
+      fork_processes if options[:processes] && options[:processes] > 1
+
+      Workerholic.manager = Manager.new(auto_balance: options[:auto_balance])
+      Workerholic.manager.start
+    end
+
+    def self.fork_processes
+      (options[:processes] - 1).times do
+        PIDS << fork do
+          Workerholic.manager = Manager.new(auto_balance: options[:auto_balance])
+          Workerholic.manager.start
+        end
+      end
+
+      PIDS.freeze
+    end
+
+    # ...
+  end
+end
+```
+
+We fork processes if the user specified the option when starting Workerholic. Then we fork the number of processes specified minus one since the main process will serve as one of these processes, effectively having 1 parent process and N - 1 children processes. We make sure to store the processes ids `pid`, inside a `PIDS` constant for future reference. In each of these children processes we start the manager, which starts all the workers and the other components such as the Job Scheduler. Once these children processes are started we start the manager in the parent process.
+
+In the next section we will look at how forking children processes impacts the consumption of the available memory.
```

Comment:
In the next section we will look at how forking child processes impacts memory usage.